src/pages/main/collaboration/reducer/reducer.ts

```diff
@@ -0,0 +1,108 @@
+import { 
+  ClientDisconnected_Message,
+  CollabReducerAction,
+  CollabState,
+  Connected_Message,
+  CursorMove_Message,
+  MessageType,
+  NewClient_Message,
+  SourceChange_Message,
+} from '../model';
+import { cloneDeep } from 'lodash';
+
+const connected_action = (state: CollabState, message: Connected_Message): CollabState => {
+  return {
+    ...state,
+    document: message.document,
+    clientId: message.clientId,
+    clientsConnectedIds: message.clientsConnectedIds,
+    cursorsPositions: message.cursorsPositions
+  };
+};
+
+const newClient_action = (state: CollabState, message: NewClient_Message): CollabState => {
+  const cursorsPositions = new Map(state.cursorsPositions);
+  cursorsPositions.set(message.clientId, null);
+
+  return {
+    ...state,
+    clientsConnectedIds: [ ...state.clientsConnectedIds, message.clientId ],
+    cursorsPositions
+  };
+};
+
+const clientDisconnected_action = (state: CollabState, message: ClientDisconnected_Message): CollabState => {
+  const cursorsPositions = new Map(state.cursorsPositions);
+  cursorsPositions.delete(message.clientId);
+  
+  return {
+    ...state,
+    clientsConnectedIds: 
+      state.clientsConnectedIds
+        .filter((clientId: string) => clientId !== message.clientId),
+    cursorsPositions
+  };
+};
+
+const sourceChange_action = (state: CollabState, message: SourceChange_Message): CollabState => {
+  return {
+    ...state,
+    deltaQueue: [ ...state.deltaQueue, { ...message } ]
+  };
+};
+
+const popDeltaQueue_action = (state: CollabState, popCount: number): CollabState => {
+  const deltaQueue = cloneDeep(state.deltaQueue);
+  deltaQueue.splice(0, popCount);
```

Comment:
Maybe use `deltaQueue.shift()`, seems like its designed for this

Reply:
very nice

Reply:
changed